src/controllers/PartialUserFormController.php

```diff
@@ -26,80 +30,120 @@ class PartialUserFormController extends UserDefinedFormController
     private static $url_handlers = [
         '$Key/$Token' => 'partial',
     ];
-
     /**
      * @var array
      */
     private static $allowed_actions = [
         'partial',
     ];
+    /**
+     * @var PartialFormSubmission
+     */
+    protected $partialFormSubmission;
 
     /**
      * Partial form
      *
      * @param HTTPRequest $request
-     * @return DBHTMLText|void
+     * @return HTTPResponse|DBHTMLText|void
      * @throws HTTPResponse_Exception
      * @throws Exception
      */
     public function partial(HTTPRequest $request)
     {
-        // Ensure this URL doesn't get picked up by HTTP caches
-        HTTPCacheControlMiddleware::singleton()->disableCache();
-
-        $key = $request->param('Key');
-        $token = $request->param('Token');
-        if (!$key || !$token) {
-            return $this->httpError(404);
-        }
-
         /** @var PartialFormSubmission $partial */
-        $partial = PartialFormSubmission::get()->find('Token', $token);
-        if (!$partial ||
-            !$partial->UserDefinedFormID ||
-            !hash_equals($partial->generateKey($token), $key)
+        $partial = $this->setData($request);
+        if ($this->dataRecord->PasswordProtected &&
```

Comment:
can we use `$this->data()` here instead of dataRecord?

Reply:
Ehhh... no, because I named the other method setData(), I'll fix up both :D